File: reference/Commands.cs
using printer;

public static class Commands
{
    // Request status: ESC i S
    public static byte[] RequestStatus = [0x1B, (byte)'i', (byte)'S'];

    // Initialize: ESC @
    public static byte[] Initialize = [0x1B, (byte)'@'];

    // Switch to Raster mode: ESC i a 01
    public static byte[] SwitchToRasterMode = [0x1B, (byte)'i', (byte)'a', 0x01];

    // Print Information: ESC i z {n1..n10}
    public static byte[] PrintInformation(PrinterStatus status, int lines)
    {
        bool isDieCut = status.MediaType == MediaType.DieCut;

        byte n1 = 0;
        n1 |= 0x80; // printer recovery always on (recommended)
        n1 |= 0x02; // KIND valid
        n1 |= 0x04; // WIDTH valid
        if (isDieCut)
        {
            n1 |= 0x08; // LENGTH valid ONLY for die-cut
        }

        byte n2 = (byte)status.MediaType; // 0x0A continuous, 0x0B die-cut
        byte n3 = status.MediaWidthMm; // width in mm
        byte n4 = isDieCut ? status.MediaLengthMm : (byte)0; // continuous: 0

        return
        [
            0x1B,
            (byte)'i',
            (byte)'z',
            n1,
            n2,
            n3,
            n4,
            (byte)(lines & 0xFF),
            (byte)((lines >> 8) & 0xFF),
            (byte)((lines >> 16) & 0xFF),
            (byte)((lines >> 24) & 0xFF), // n5..n8: raster count (0=unspecified over USB)
            0x00,
            0x00, // n9,n10
        ];
    }

    // Auto Cut: ESC i M @
    public static byte[] AutoCut = [0x1B, (byte)'i', (byte)'M', (byte)'@'];

    // Set to 600dpi + cut at end: ESC i K H
    public static byte[] Set600DpiAndCut = [0x1B, (byte)'i', (byte)'K', (byte)'H'];

    // Set to 600dpi + cut at end: ESC i K H
    public static byte[] CutAtEnd = [0x1B, (byte)'i', (byte)'K', 0x08];

    // Set margin amount: ESC i d 0x23 0x00
    public static byte[] SetMarginAmount = [0x1B, (byte)'i', (byte)'d', 0x23, 0x00];

    // Raster data: g 0x00 0x5A {byte[] rasterData}
    public static byte[] RasterData(byte[] rasterData)
    {
        // MaskMargins62mm(rasterData);
        var length = rasterData.Length;
        var command = new byte[3 + length];
        command[0] = (byte)'g';
        command[1] = 0x00; // Raster data type
        command[2] = 0x5A; // 90 bytes of raster data
        Array.Copy(rasterData, 0, command, 3, length);
        return command;
    }

    // End of page: 0x0C
    public static byte[] EndOfPage = [0x0C];

    // Print With Feeding: 0x1A
    public static byte[] PrintWithFeeding = [0x1A];

    // No Compression: M 0x00
    public static byte[] NoCompression = [0x4D, 0x00];

    // Automatic Status Notification: ESC i ! 0x00
    public static byte[] AutomaticStatusNotification = [0x1B, (byte)'i', (byte)'!', 0x00];

    // Cut each 1: ESC i A 0x01
    public static byte[] CutEach1 = [0x1B, (byte)'i', (byte)'A', 0x01];

    static void MaskMargins62mm(
        byte[] line /* len 90 */
    )
    {
        // Left 12 bits: byte0 all 0, top 4 bits of byte1 0
        line[0] = 0x00;
        line[1] = 0x00;
        line[2] = 0x00;
        line[3] &= 0x0F; // keep only low 4 bits

        // Right 12 bits: bottom 4 bits of byte88 0, byte89 all 0
        line[86] &= 0xF0; // keep only high 4 bits
        line[87] = 0x00;
        line[88] = 0x00;
        line[89] = 0x00;
    }
}

File: packages/brother-ql-web/src/core/commands.ts
// Brother Raster commands (ported from reference Commands.cs)
// All values are in hex; using Uint8Array for transferOut

import type { PrinterStatus } from "./status";
import { MEDIA_TYPE } from "./status";

export const RequestStatus = new Uint8Array([0x1b, 0x69, 0x53]); // ESC i S
export const Initialize = new Uint8Array([0x1b, 0x40]); // ESC @
export const SwitchToRasterMode = new Uint8Array([0x1b, 0x69, 0x61, 0x01]); // ESC i a 01
export const AutoCut = new Uint8Array([0x1b, 0x69, 0x4d, 0x40]); // ESC i M @
export const CutAtEnd = new Uint8Array([0x1b, 0x69, 0x4b, 0x08]); // ESC i K 08
export const SetMarginAmount = new Uint8Array([0x1b, 0x69, 0x64, 0x23, 0x00]); // ESC i d 0x23 0x00
export const EndOfPage = new Uint8Array([0x0c]);
export const PrintWithFeeding = new Uint8Array([0x1a]);
export const NoCompression = new Uint8Array([0x4d, 0x00]); // M 00
export const AutomaticStatusNotification = new Uint8Array([0x1b, 0x69, 0x21, 0x00]); // ESC i ! 00
export const CutEach1 = new Uint8Array([0x1b, 0x69, 0x41, 0x01]); // ESC i A 01

export function printInformation(status: PrinterStatus, lines: number): Uint8Array {
  const isDieCut = status.mediaType === MEDIA_TYPE.DieCut;

  let n1 = 0;
  n1 |= 0x80; // printer recovery always on (recommended)
  n1 |= 0x02; // KIND valid
  n1 |= 0x04; // WIDTH valid
  if (isDieCut) n1 |= 0x08; // LENGTH valid ONLY for die-cut

  const n2 = status.mediaType & 0xff; // media type
  const n3 = status.mediaWidthMm & 0xff; // width in mm
  const n4 = isDieCut ? status.mediaLengthMm & 0xff : 0x00; // continuous: 0

  return new Uint8Array([
    0x1b,
    0x69,
    0x7a, // 'z'
    n1,
    n2,
    n3,
    n4,
    lines & 0xff,
    (lines >> 8) & 0xff,
    (lines >> 16) & 0xff,
    (lines >> 24) & 0xff, // n5..n8: raster count
    0x00,
    0x00, // n9, n10
  ]);
}

export function rasterData(line: Uint8Array): Uint8Array {
  const cmd = new Uint8Array(3 + line.length);
  cmd[0] = 0x67; // 'g'
  cmd[1] = 0x00; // raster data type
  cmd[2] = 0x5a; // 90 bytes following
  cmd.set(line, 3);
  return cmd;
}
